src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -0,0 +1,137 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    // NOTE: this type depends on Dev15 assemblies, which is why the type is in EditorFeatures.Next. But, that library
+    // is rehostable and once we move .editorconfig support fully through the system, it should be moved to Workspaces
+    // or perhaps even lower.
+    internal sealed class EditorConfigDocumentOptionsProvider : IDocumentOptionsProvider
+    {
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// The map of cached contexts for currently open documents.
+        /// </summary>
+        private readonly Dictionary<DocumentId, Task<ICodingConventionContext>> _openDocumentContexts = new Dictionary<DocumentId, Task<ICodingConventionContext>>();
+
+        private readonly ICodingConventionsManager _codingConventionsManager;
+
+        internal EditorConfigDocumentOptionsProvider(Workspace workspace)
+        {
+            _codingConventionsManager = CodingConventionsManagerFactory.CreateCodingConventionsManager();
+
+            workspace.DocumentOpened += Workspace_DocumentOpened;
+            workspace.DocumentClosed += Workspace_DocumentClosed;
+        }
+
+        private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
+        {
+            lock (_gate)
+            {
+                Task<ICodingConventionContext> contextTask;
+
+                if (_openDocumentContexts.TryGetValue(e.Document.Id, out contextTask))
+                {
+                    _openDocumentContexts.Remove(e.Document.Id);
+
+                    // Ensure we dispose the context, which we'll do asynchronously
+                    contextTask.ContinueWith(t => t.Result.Dispose(), CancellationToken.None, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+                }
+            }
+        }
+
+        private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
+        {
+            lock (_gate)
+            {
+                _openDocumentContexts.Add(e.Document.Id, Task.Run(() => _codingConventionsManager.GetConventionContextAsync(e.Document.FilePath, CancellationToken.None)));
+            }
+        }
+
+        public async Task<IDocumentOptions> GetOptionsForDocumentAsync(Document document, CancellationToken cancellationToken)
+        {
+            Task<ICodingConventionContext> contextTask;
+
+            lock (_gate)
+            {
+                _openDocumentContexts.TryGetValue(document.Id, out contextTask);
+            }
+
+            if (contextTask != null)
+            {
+                // The file is open, let's reuse our cached data for that file. The task might still be running, but we want to allow for eager cancellation
+                // if the caller doesn't need it.
+                var cancellableContextTask = contextTask.ContinueWith(
```

Comment:
the comment isn't really clear.  I think you want to explain that 'awaiting' the contextTask means that we won't respect 
cancellation.  So we do this so we can both get the underlying value and respect cancellation if the caller wants that.